### DISCUSSION TITLE
SofaCaribou compilation on Windows

### DISCUSSION THREAD

**Discussion by ruiliang-gao** [jnbrunet/caribou#50] [Q&A] [answered]
I am having cmake errors trying to compile SofaCairbou on sofa v20.12 and latest Caribou source.
I added the SOFA_INSTALL environment variable as stated in the instructions. 
My cmake version is 3.20.1, compiler specified for x64, VS2019.
I set "Caribou_DIR" to "CaribouSrc/caribou/src/Caribou" which is where I clone the Caribou source code.

The full Cmake output is attached below:
        
        Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19042.
        Install prefix: C:/Users/Ruiliang/Documents/sofamaster/build/install
        CMake Deprecation Warning at CMakeLists.txt:222 (message):
          Deprecated components are activated (SOFA_WITH_DEPRECATED_COMPONENTS
          variable is ON)
        
       
        Adding module SofaSimpleFem
        Adding module SofaRigid
        Adding module SofaDeformable
        Adding module SofaObjectInteraction
        Adding module SofaMeshCollision
        Adding module SofaEngine
        Adding module SofaExplicitOdeSolver
        Adding module SofaImplicitOdeSolver
        Adding module SofaLoader
        Adding module SofaGeneralVisual
        Adding module SofaGraphComponent
        Adding module SofaGeneralMeshCollision
        Adding module SofaBoundaryCondition
        Adding module SofaGeneralAnimationLoop
        Adding module SofaGeneralDeformable
        Adding module SofaGeneralEngine
        Adding module SofaGeneralExplicitOdeSolver
        Adding module SofaGeneralImplicitOdeSolver
        Adding module SofaGeneralLinearSolver
        Adding module SofaGeneralRigid
        Adding module SofaGeneralObjectInteraction
        Adding module SofaGeneralSimpleFem
        Adding module SofaGeneralTopology
        Adding module SofaTopologyMapping
        Adding module SofaUserInteraction
        Adding module SofaConstraint
        Adding module SofaGeneralLoader
        Adding module SofaSparseSolver
        SofaSparseSolver: using built-in metis library
        Adding module SofaPreconditioner
        Adding module SofaHaptics
        Adding module SofaValidation
        Adding module SofaDenseSolver
        Adding module SofaNonUniformFem
        Adding module SofaOpenglVisual
        Adding module SofaMiscTopology
        Adding module SofaMiscExtra
        Adding module SofaMiscForceField
        Adding module SofaMiscEngine
        Adding module SofaMiscSolver
        Adding module SofaMiscFem
        Adding module SofaMiscMapping
        Adding module SofaExporter
        SofaPython disabled, will not compile python bindings.
        Adding collection SofaComponentAll
        CMake Deprecation Warning at C:/Users/Ruiliang/Documents/sofamaster/build/cmake/SofaCommonConfig.cmake:33 (message):
           SofaCommon is deprecated.
           You must now explicitely find (find_package) and link (target_link_library) the modules you need within: 
           SofaSimpleFem;SofaRigid;SofaDeformable;SofaObjectInteraction;SofaMeshCollision;SofaEngine;SofaExplicitOdeSolver;SofaImplicitOdeSolver;SofaLoader
        Call Stack (most recent call first):
          applications/collections/SofaComponentAll/CMakeLists.txt:5 (find_package)
        
        
        CMake Deprecation Warning at C:/Users/Ruiliang/Documents/sofamaster/build/cmake/SofaGeneralConfig.cmake:33 (message):
           SofaGeneral is deprecated.
           You must now explicitely find (find_package) and link (target_link_library) the modules you need within: 
           SofaBoundaryCondition;SofaGeneralMeshCollision;SofaGeneralVisual;SofaGraphComponent;SofaGeneralAnimationLoop;SofaGeneralDeformable;SofaGeneralEngine;SofaGeneralExplicitOdeSolver;SofaGeneralImplicitOdeSolver;SofaGeneralLinearSolver;SofaGeneralRigid;SofaGeneralObjectInteraction;SofaGeneralSimpleFem;SofaGeneralTopology;SofaTopologyMapping;SofaUserInteraction;SofaConstraint;SofaGeneralLoader
        Call Stack (most recent call first):
          applications/collections/SofaComponentAll/CMakeLists.txt:6 (find_package)
        
        
        CMake Deprecation Warning at C:/Users/Ruiliang/Documents/sofamaster/build/cmake/SofaMiscConfig.cmake:33 (message):
           SofaMisc is deprecated.
           You must now explicitely find (find_package) and link (target_link_library) the modules you need within: 
           SofaMiscExtra;SofaMiscEngine;SofaMiscFem;SofaMiscForceField;SofaMiscMapping;SofaMiscSolver;SofaMiscTopology
        Call Stack (most recent call first):
          applications/collections/SofaComponentAll/CMakeLists.txt:7 (find_package)
        
        
        Adding plugin CImgPlugin
        CImgPlugin: TIFF Image Format enabled
        CImgPlugin: JPEG Image Format enabled
        CImgPlugin: PNG Image Format enabled
        Adding plugin SofaMiscCollision
        SofaMiscModule: build all maintained sofa components.
        SofaMiscCollision: optional dependency SofaSphFluid not found de-activing: <SpatialGridPointModel>
        Adding plugin SofaCaribou
        CMake Error at applications/plugins/SofaCaribou/CMakeLists.txt:97 (find_package):
          By not providing "FindCaribou.cmake" in CMAKE_MODULE_PATH this project has
          asked CMake to find a package configuration file provided by "Caribou", but
          CMake did not find one.
        
          Could not find a package configuration file provided by "Caribou" with any
          of the following names:
        
            CaribouConfig.cmake
            caribou-config.cmake
        
          Add the installation prefix of "Caribou" to CMAKE_PREFIX_PATH or set
          "Caribou_DIR" to a directory containing one of the above files.  If
          "Caribou" provides a separate development package or SDK, be sure it has
          been installed.

**Comment by jnbrunet** [comment, marked as answer]
Hi @rgao15 ,

You don't need to set the `Caribou_DIR` cmake variable. Can you try the following:

1. Create a directory somewhere on your PC named `sofa_plugins`. Clone the source code of Caribou inside of it. For example:
```
$ mkdir "C:/Users/Ruiliang/Documents/sofa_plugins"
$ git clone https://github.com/jnbrunet/caribou.git "C:/Users/Ruiliang/Documents/sofa_plugins/caribou"
```
2. Create a `C:/Users/Ruiliang/Documents/sofa_plugins/CMakeLists.txt` file containing the following:
```cmake
cmake_minimum_required(VERSION 3.12)

find_package(SofaFramework)
sofa_add_plugin(caribou/ SofaCaribou)
```
3. Go into your SOFA's build directory, and add this `sofa_plugins` path to the `SOFA_EXTERNAL_DIRECTORIES` CMake variable. Then, activate the compilation of the plugin using the CMake variable `PLUGIN_SOFACARIBOU`. For example:
```
$ cd C:/Users/Ruiliang/Documents/sofamaster/build
$ cmake -DSOFA_EXTERNAL_DIRECTORIES="C:/Users/Ruiliang/Documents/sofa_plugins" .
$ cmake -DPLUGIN_SOFACARIBOU=ON .
```

Finally, build SOFA. This should build the SofaCaribou plugin at the same time.

**Reply by ruiliang-gao** [reply]
Thank you, i just got it compiled with your solutions! will start playing with Caribou.

**Reply by jnbrunet** [reply]
Awesome, let me know if you hit other issues.
